### PR TITLE
support ELEMENTALSETS in FEFLOW import

### PR DESCRIPTION
**Pull request by norihiro-w**
This PR supports the keyword ELEMENTALSETS added in recent FEFLOW version to set material IDs to imported elements. joint work with @fabienma

### CODE REVIEW

File: FileIO/FEFLOWInterface.cpp [hunk 1]
```diff
@@ -139,6 +141,54 @@ MeshLib::Mesh* FEFLOWInterface::readFEFLOWFile(const std::string &filename)
 			line_stream.clear();
 		}
 		//....................................................................
+		// ELEMENTALSETS
+		else if (line_string.compare("ELEMENTALSETS") == 0)
+		{
+			unsigned mode = 0;
+			std::string str_nodeList;
+			int pos_prev_line = 0;
+			while (true)
```
Comment (endJunction) [inline FileIO/FEFLOWInterface.cpp]: Could be `do { ... } while (mode != 0)`, or change the `mode = 0` lines below to `break` and make the mode variable local to the loop.

File: FileIO/FEFLOWInterface.cpp [hunk 2]
```diff
@@ -139,6 +141,54 @@ MeshLib::Mesh* FEFLOWInterface::readFEFLOWFile(const std::string &filename)
...
+			{
+				pos_prev_line = in.tellg();
```
Comment (endJunction) [inline FileIO/FEFLOWInterface.cpp]: `tellg()` returns a `long`;

File: FileIO/FEFLOWInterface.cpp [hunk 3]
```diff
@@ -460,10 +523,15 @@ void FEFLOWInterface::readSuperMesh(std::ifstream &in, const FEM_CLASS &fem_clas
...
-void FEFLOWInterface::setMaterialID(const FEM_CLASS &fem_class, const FEM_DIM &fem_dim, const std::vector<GeoLib::Polyline*>* lines, std::vector<MeshLib::Element*> &vec_elements)
+void FEFLOWInterface::setMaterialID(const FEM_CLASS &fem_class, const FEM_DIM &fem_dim, const std::vector<GeoLib::Polyline*>* lines, const std::vector<std::vector<std::size_t>> &vec_elementsets, std::vector<MeshLib::Element*> &vec_elements)
 {
-	if (lines && lines->size() > 1)
-	{
+	if (!vec_elementsets.empty()) {
+		for (std::size_t matid =0; matid<vec_elementsets.size(); matid++) {
+			auto &eids = vec_elementsets[matid];
+			for (auto eid : eids)
+				vec_elements[eid-1]->setValue(matid); // Element IDs given by FEFLOW starts from one!
```
Comment (endJunction) [inline FileIO/FEFLOWInterface.cpp]: FYI: setValue takes an unsigned int, losing integer precision implicitly.
Should be handled by changing setValue's parameter to a std::size_t.

**Comment by endJunction** [comment]
:fast_forward: even w/o changes.

**Comment by bilke** [comment]
:+1:

**Comment by norihiro-w** [comment]
found some bug. needs to be fixed

**Comment by norihiro-w** [comment]
the bug fix is done

**Comment by fabienma** [comment]
Thank you! I will test it.

**Comment by norihiro-w** [comment]
@fabienma fixed another bug. now you can import small_cube_hex.fem

**Comment by fabienma** [comment]
Thank you Nori, I will check it soon!

On 8/5/2015 9:40 PM, Norihiro Watanabe wrote:
>
> @fabienma <https://github.com/fabienma> fixed another bug. now you can 
> import small_cube_hex.fem
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/ufz/ogs/pull/794#issuecomment-128124868>.
>

-- 
PD Dr. Fabien Magri

Department of Environmental Informatics (ENVINF)
Helmholtz-Zentrum für Umweltforschung GmbH - UFZ
Permoserstraße 15 I 04318 Leipzig
Tel.: +49 341 235 - 1059
Fax: +49 341 235 - 451059
Email: fabien.magri@ufz.de
WWW: http://www.ufz.de

Helmholtz-Zentrum für Umweltforschung GmbH - UFZ
Sitz der Gesellschaft: Leipzig
Registergericht: Amtsgericht Leipzig
Handelsregister Nr. B 4703
Vorsitzender des Aufsichtsrats: MinDirig Wilfried Kraus
Wissenschaftlicher Geschäftsführer: Prof. Dr. Georg Teutsch
Administrative Geschäftsführerin: Dr. Heike Graßmann

**Pull request merged**